Member정보관련 유효성 검사 추가#39

nahokyun opened the pull request:
> ### 추가 및 변경된점
- validation사용을위해 org.springframework.boot:spring-boot-starter-validation를 추가하였습니다.
- 회원가입에 실패했을때의 message를 ExceptionMessageConst에 추가하였습니다.
- validation에 걸렸을때 발생하는 exception을 handling하였습니다.
## MemberController.java
### 변경 전 
```java
    public String memberJoin(@RequestBody MemberDTO memberDTO) {
```
### 변경 후 
```java
    public String memberJoin(@RequestBody @Valid MemberDTO memberDTO) {
```
## MemberDTO.java
### 변경 전 
```java
    private final String email;

    private final String loginId;

    private final String loginPassword;

    private final String nickname;

```
### 변경 후 
```java
    @NotBlank
    @Pattern(regexp = "^[A-Za-z0-9_\\.\\-]+@[A-Za-z0-9\\-]+\\.[A-Za-z0-9\\-]+$")
    private final String email;

    @NotBlank
    private final String loginId;

    @NotBlank
    @Size(min=2,max=10)
    private final String loginPassword;

    @NotBlank
    private final String nickname;

```
## GeneralExceptionHandler.java
### 추가된 메소드
```java
    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public String methodArgumentNotValidHandler(MethodArgumentNotValidException e){
        return ExceptionMessageConst.FAILED_SIGN_UP.getMessage();
    }

```

nahokyun committed:
Member정보관련 유효성 검사 추가#39

validation사용을위해 org.springframework.boot:spring-boot-starter-validation를 추가하였습니다.

회원가입에 실패했을때의 message를 추가하였습니다.

validation에 걸렸을때 발생하는 exception을 handling하였습니다.

sunwootest commented on src/main/java/study/movieservice/domain/member/MemberDTO.java:
```diff
@@ -3,11 +3,25 @@
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.Size;
+
 @Getter
 @RequiredArgsConstructor
 public class MemberDTO {
+    @NotBlank
+    @Pattern(regexp = "^[A-Za-z0-9_\\.\\-]+@[A-Za-z0-9\\-]+\\.[A-Za-z0-9\\-]+$")
     private final String email;
+
+    @NotBlank
     private final String loginId;
+
+    @NotBlank
+    @Size(min=2,max=10)
     private final String loginPassword;
+
+    @NotBlank
     private final String nickname;
+
```
loginId와 nickname의 검증은 이정도면 충분한가요?

nahokyun replied:
loginId는 영문 숫자조합으로 4글자 이상 10글자 미만으로 처리하는 정도로 수정하면 충분할거라 생각되고 nickname의 경우는 최소글자만 지정해두는 정도로 처리하면 충분할것같습니다.

sunwootest commented on src/main/java/study/movieservice/domain/member/MemberDTO.java:
```diff
@@ -3,11 +3,25 @@
 @Getter
 @RequiredArgsConstructor
 public class MemberDTO {
+    @NotBlank
+    @Pattern(regexp = "^[A-Za-z0-9_\\.\\-]+@[A-Za-z0-9\\-]+\\.[A-Za-z0-9\\-]+$")
     private final String email;
```
\@email 이란 것도 있습니다 :)

nahokyun replied:
@email의 경우 찾아보니 빈 문자열("")이 통과가 된다고 하여 정규표현식으로 사용하였는데 지금 생각해보니 빈 문자열을 검증해주는 @NotNull과 같이 사용하면 해결이 가능할것 같습니다!

nahokyun replied:
방금 작성해보니 NotNull을 사용해도 빈 문자열이 성공한다고 처리되어 @Email 사용은 좀더 알아본뒤에 적용하도록 하겠습니다.

nahokyun committed:
Member정보관련 유효성 검사 추가#39

loginId와 nickname에 대한 유효성 조건을 추가하였습니다.

nahokyun commented:
> ### 변경된점
- loginId와 nickname에 대한 유효성 조건을 추가하였습니다.
## MemberDTO.java
### 변경 전 
```java
    @NotBlank
    private final String loginId;

    @NotBlank
    private final String nickname;
```
### 변경 후 
```java
    @NotBlank
    @Pattern(regexp = "^[A-Za-z0-9]{4,9}")
    private final String loginId;

    @NotBlank
    @Size(min = 2)
    private final String nickname;
```

nahokyun committed:
Member정보관련 유효성 검사 추가#39

MemberService 클래스에 Validated어노테이션과 addMember 메소드에 @Valid를 사용함으로 service단계에서 MemberDTO에대한 유효성 검사를 진행할수 있도록 추가하였습니다.

nahokyun commented:
> ### 변경된점
- MemberService 단계에서도 MemberDTO에대한 유효성 검사를 진행하도록 추가하였습니다.
- @Validated를 사용했을때 발생하는 ConstraintViolationException 핸들링에관한 내용을 추가하였습니다. 
- ExceptionMessageConst.java에 Service단계에서 유효성검사를 실패했을때 발생하는 FAILED_DATA_TRANSMISSION_SERVICE를 추가하였습니다.
## MemberService.java
### 변경 전 
```java
@Service
@RequiredArgsConstructor
public class MemberService {

    private final MemberMapper memberMapper;
    private final MailService mailService;
    private final SessionManager sessionManager;

    public void addMember(MemberDTO memberDTO){

```
### 변경 후 
```java
@Service
@RequiredArgsConstructor
@Validated
public class MemberService {

    private final MemberMapper memberMapper;
    private final MailService mailService;
    private final SessionManager sessionManager;

    public void addMember(@Valid MemberDTO memberDTO){

```
## GeneralExceptionHandler.java
### 추가된 메소드 
```java
    @ExceptionHandler(ConstraintViolationException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public String constraintViolationHandler(ConstraintViolationException e){
        return ExceptionMessageConst.FAILED_SIGN_UP.getMessage();
    }

```

sunwootest reviewed (approved):
approve 드립니다!
